# -*- coding: utf-8 -*-
"""
Created on Wed Sep 18 15:33:06 2024

@author: 67535
"""

import vcf
import pandas as pd
import argparse

parser = argparse.ArgumentParser(description='Use the ARO list of antibiotics of interested to search for the sequences of antibiotic resistance genes (ARGs).')

parser.add_argument('-o', '--output', required = False,
                    help='Output fasta file path and name. Default: ./output/screened_primers_probes.xlsx',default='./output/screened_primers_probes.xlsx')
parser.add_argument('-v', '--vcf',required = True,
                    help='Input vcf file path generated by FreeBayes for the ARG mutations detected in the given sample.')
parser.add_argument('-f', '--fasta', required = True,
                    help='The fasta file generated by "drug_to_corresponding_ARGs_using_owl.py".')
parser.add_argument('-p', '--primer',required = True,
                    help='The designed primers and probes downloaded from IDT PrimerQuest Tool.')
args = parser.parse_args()

#Read and process VCF file from FreeBayes
#################################################################################################################
vcf_reader = vcf.Reader(filename=args.vcf)
#################################################################################################################
variant_table=pd.DataFrame(columns=['ARO','POS','REF','ALT','QUAL','DP'])

for record in vcf_reader:  
    if record.QUAL >=20:
        variant_record=pd.DataFrame(columns=['ARO','POS','REF','ALT','QUAL','DP'],data=[[record.CHROM,record.POS,record.REF,record.ALT,record.QUAL,record.INFO['DP']]])
        variant_table=pd.concat([variant_table,variant_record],ignore_index=True)
        
#Read the reference fasta for primer and probe design
##############################################################################################
file=open(args.fasta)
##############################################################################################
ref_fasta=file.readlines()
file.close()

ref_table=pd.DataFrame(columns=['ARO','SEQ'])
for i in range(0,len(ref_fasta),2):
    ref_record=pd.DataFrame(columns=['ARO','SEQ'],data=[[ref_fasta[i].strip('>\n'),ref_fasta[i+1].strip('\n')]])
    ref_table=pd.concat([ref_table,ref_record],ignore_index=True)

#Read and process IDT primer set output: get the regions that the primers and probes covered
###################################################################################################################
raw_primers=pd.read_excel(args.primer)
###################################################################################################################
primer_sites=pd.DataFrame(columns=['ARO','AssaySet','Forward','Reverse','Probe','Sites'])
for i in range(0,len(raw_primers),4):
    ARO=raw_primers.loc[i,'AssaySet'].split('(')[1].split(')')[0]
    sites=list(range(int(raw_primers.loc[i,'Start']),int(raw_primers.loc[i,'Start'])+int(raw_primers.loc[i,'Length'])))
    sites=sites+list(range(int(raw_primers.loc[i+2,'Start'])-int(raw_primers.loc[i+2,'Length'])+1,int(raw_primers.loc[i+2,'Start'])+1))
    print(i)
    if raw_primers.loc[i+1,'Sequence']==str(list(ref_table[ref_table['ARO']==ARO]['SEQ'])).strip("[']")[int(raw_primers.loc[i+1,'Start'])-1:int(raw_primers.loc[i+1,'Start'])+int(raw_primers.loc[i+1,'Length'])-1]:
        sites=sites+list(range(int(raw_primers.loc[i+1,'Start']),int(raw_primers.loc[i+1,'Start'])+int(raw_primers.loc[i+1,'Length'])))
    else:
        sites=sites+list(range(int(raw_primers.loc[i+1,'Start'])-int(raw_primers.loc[i+1,'Length'])+1,int(raw_primers.loc[i+1,'Start'])+1))
    primer_sites_temp=pd.DataFrame(columns=['ARO','AssaySet','Forward','Reverse','Probe','Sites','ProductLength'],data=[[raw_primers.loc[i,'AssaySet'].split('(')[1].split(')')[0], raw_primers.loc[i,'AssaySet'].split('Set ')[1].split(' (')[0], raw_primers.loc[i,'Sequence'], raw_primers.loc[i+2,'Sequence'],raw_primers.loc[i+1,'Sequence'],sites,raw_primers.loc[i+3,'Amplicon']]])
    primer_sites=pd.concat([primer_sites,primer_sites_temp],ignore_index=True)
    
#According to AROs in ref_fasta, find out the mutated regions

output_df=pd.DataFrame(columns=['ARO','AssaySet','Forward','Reverse','Probe','Length'])
for i in range(0,len(ref_table)):
    pos_list=list(variant_table[variant_table['ARO']==ref_table.loc[i,'ARO']]['POS'])
    ref_list=list(variant_table[variant_table['ARO']==ref_table.loc[i,'ARO']]['REF'])
    ref_len_list=[]
    for j in range(0,len(ref_list)):
        ref_len_list.append(len(ref_list[j]))
    for j in range(0,len(ref_len_list)):
        if ref_len_list[j]!=1:
            pos_list=pos_list+list(range(pos_list[j]+1,pos_list[j]+ref_len_list[j]))
    pos_list=list(sorted(list(set(pos_list))))
    for j in range(0,len(primer_sites)):
        if primer_sites.loc[j,'ARO']==ref_table.loc[i,'ARO']:
            inter_list=list(set(pos_list).intersection(set(primer_sites.loc[j,'Sites'])))
            if inter_list==[]:
                output_temp=pd.DataFrame(columns=['ARO','AssaySet','Forward','Reverse','Probe','Length'],data=[[primer_sites.loc[j,'ARO'],primer_sites.loc[j,'AssaySet'],primer_sites.loc[j,'Forward'],primer_sites.loc[j,'Reverse'],primer_sites.loc[j,'Probe'],primer_sites.loc[j,'ProductLength']]])
                output_df=pd.concat([output_df,output_temp],ignore_index=True)

#Write output
################################################################################################################
output_df.to_excel(args.output)
################################################################################################################
